channeldb/migration32/mission_control_store.go

```diff
@@ -163,7 +163,14 @@ func newPaymentFailure(sourceIdx *int,
 		sourceIdx: tlv.NewPrimitiveRecord[tlv.TlvType0](
 			uint8(*sourceIdx),
 		),
-		msg: tlv.NewRecordT[tlv.TlvType1](failureMessage{failureMsg}),
+	}
+
+	if failureMsg != nil {
```

Comment:
Add a comment why this might happen, decoding error of the error msg.

What about the case where we have a nil source index:

https://github.com/lightningnetwork/lnd/blob/b34afa33f6993e4f24e481f83757751f0f8983de/routing/missioncontrol.go#L843-L845

we should instead return nil here, so we do not create the failure information.

Reply:
> we should instead return nil here, so we do not create the failure information.

Ah, I think we should do something about the pointer, I converted to use a `fn.Option[paymentFailure]` in other places and removed the pointer, because `newPaymentFailure` always constructs a `paymentFailure`, it still represents a payment failure if the info fields are absent.

Reply:
> Add a comment why this might happen, decoding error of the error msg.

I added comments to the `paymentFailure` fields.

> What about the case where we have a nil source index

The `sourceIdx` is a prerequisite to be able to decrypt the error via the hop's shared secret, so I think it makes sense to return an empty `paymentFailure` here, because if we don't know the index, we won't be able to get a message.

Reply:
I more or less had the same question when looking at the code. Perhaps this explanation is worth a comment in both versions of `newPaymentFailure` where we return an empty `paymentFailure`.

Reply:
It looks like one level of nesting can be removed, trying that out.

Reply:
It seems to work, I removed `paymentFailure`, whose role is now taken over by the old `paymentFailureInfo`. Instead of checking the optional record in `paymentFailure`, I think we can use the source index instead. I put all mission control migration changes together in the last commit, which was easier for me. Let me know if I should make changes in lockstep.